sycl/include/CL/sycl/context.hpp

```diff
@@ -133,10 +135,21 @@ class context {
   /// \return a vector of valid SYCL device instances.
   vector_class<device> get_devices() const;
 
+  /// Gets the native handle of the SYCL context.
+  ///
+  /// \return a native handle, the type of which defined by the backend.
+  template <backend BackendName>
+  auto get_native() const -> typename interop<BackendName, context>::type {
+    return reinterpret_cast<typename interop<BackendName, context>::type>(
```

Comment:
We have to restore the type here because we have on PI call which is supposed to cast any handle?

Reply:
The reason we do the cast here is because we have the `Backend` template parameter which determines the handle type, so the definition of `get_native` must be in the header. However, since `context_impl` is forward declared here we cannot call any of it's methods, thus we have the `getNative` private method on the `context` with it's definition in context.cpp. This method calls into the impl which calls the PI API using the corresponding plugin. All the way through we use the general `pi_native_handle` type as the native handle is only known based on the backend template parameter that is only known by `get_native`.

Does this answer your question?

Reply:
We don't have many backends: only 3, so if we add specializations per each would it eliminate the need in pi_native_handle here?

Reply:
Where would these specializations be? Correct me if I'm wrong, but if it's in `context.cpp` we would have to include the backend headers as the return type of the specializations of `get_native` would be fully resolved. If it's mixed in with the other backend specializations we have to put it it the headers as we do not have source files for these, so we're as well off as with this.